Moved to version 2.2.0 and added windows support

willyd opened the pull request:
gflags for windows.

For some reason testing for the existence of the dll fails the build (see https://github.com/willyd/gflags-feedstock/blob/685e45515183adba6e4303117d2761b1902e4b37/recipe/meta.yaml#L31-L34). Is there anything wrong in my meta.yaml?

willyd committed:
Moved to version 2.2.0 and added windows support

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found some lint.

Here's what I've got...


For **recipe**:

 * Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form.

willyd committed:
Fixed lint issue

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

jakirkham commented:
Neat! Thanks for putting this together.

Could you please [re-render]( https://github.com/conda-forge/conda-smithy#re-rendering-an-existing-feedstock ) with `conda-smithy` version `1.7.0`? This will enable builds on AppVeyor.

Also make sure before doing the re-rendering that your `.condarc` includes a channel section that looks like the one below. After re-rendering you can remove it. In future `conda-smithy` releases, that shouldn't be necessary.

```yaml
channels:
  - conda-forge
  - defaults
```

jakirkham commented on recipe/meta.yaml:
```diff
@@ -7,22 +7,31 @@ package:
 
 build:
   number: 0
-  skip: true  # [win]
```
Please keep, but change like so.

```yaml
  skip: true  # [win and py==36]
```

Note that Python 3.5 and 3.6 use the same VC version. So there is no need to build both. 😄

jakirkham commented on recipe/meta.yaml:
```diff
@@ -7,22 +7,31 @@ package:
+    # For some reason testing for the existence of the dll fails the build with:
+    #   The following specifications were found to be in conflict:
+    #   - gflags 2.2.0 vc14_0 -> feature:vc14
+    # - if not exist %PREFIX%\\Library\\bin\\gflags.dll exit 1   # [win]
```
Please add the following above in `test`.

```yaml
  requires:
    - python 2.7.*  # [win and py27]
    - python 3.4.*  # [win and py34]
    - python 3.5.*  # [win and py35]
```

jakirkham replied:
This should fix it.

willyd commented:
Thanks for the quick review.

> Could you please re-render with conda-smithy version 1.7.0? This will enable builds on AppVeyor.

Ok. Not sure what this means but I guess your link will explain that in details. 

I will correct the above issues and update the PR.

willyd committed:
Disabled build for python 3.6 and fixed tests on windows

jakirkham commented:
`conda-smithy` is the cookie cutter of feedstocks. It fills them out with all the CI files, connects them to all the CIs, and configures them for uploading. Though it also can be useful for changing a feedstock after some recipe change (e.g. enabling/disabling builds on a platform, adding/removing other Python or NumPy versions, displaying metadata in the Readme, etc.).

wesm commented:
Can this be updated? I need to add static libs to OSX/Linux, but I can do that in a separate patch. Didn't want to cause problems with this patch

willyd committed:
Rerendered with conda-smithy 2.0.1

willyd commented:
@jakirkham @wesm I rendered with smithy 2.0.1 and applied the requested changes.

wesm commented:
Looks like you may want `mkdir -p` in the `build.sh`, or else use a name for the build directory that's more unique.

willyd commented:
I should probably disable the build for python 2.7 as gflags does not compile with VS 2008. https://github.com/gflags/gflags/issues/197

wesm commented:
Please do. We can update the build once the gflags bug is resolved

willyd committed:
Skip unsupported vc version
Renamed build folder to avoid conflict with the bazel BUILD file

willyd commented:
@wesm. I fixed the mkdir and vc9 issues. For some reason the Travis build does not start. Is this a normal thing?

wesm commented:
@willyd yes, it often takes the OS X builds a long time to run, especially during the week. I would check back every 12 hours or so

wesm commented:
@jakirkham @willyd I think we should merge this to get master into the OS X build queue, objections?

willyd commented:
None on my side.

jakirkham commented:
Travis CI appears to have frozen. Toggling to get it restarted on the tail of the weekend.

jakirkham commented on recipe/meta.yaml:
```diff
@@ -7,22 +7,36 @@ package:
   number: 0
-  skip: true  # [win]
+  # skipping
+  # py27 (vc9) and py34 (vc10) since vc does not have stdint.h
```
This should be solved by adding `msinttypes` as a `build` dependency on Windows particularly for these Pythons.

jakirkham commented:
Would you be interested in adding yourself as a maintainer, @willyd? I don't know how much time I can devote to Windows maintenance of this package myself (particularly of late). However, if you are willing to chip in, then I would be willing to consider it.

willyd commented:
@jakirkham  Sorry for the late, late response. I don't have much time to spend on this now either so I have to decline your offer. If this means that you can't accept the PR if I make the changes your requested it is understandable. Thanks for the work you are making on the conda ecosystem!

jakirkham commented on recipe/patch_cmake_targets.py:
```diff
@@ -0,0 +1,18 @@
+        fout.write(content)
+
+if __name__ == '__main__':
+    file_replace(*sys.argv[1:])
```
Basically the same story as OpenCV, it would be better to have actual patches instead of a patching script like this.

wesm commented:
@jakirkham @willyd I'm willing to help maintain the package on Windows as we are using gflags in Apache Arrow, and I'd like people to be able to `conda install gflags` as part of setting up their build toolchain. What work needs to be done to get this in?

willyd commented:
@wesm That is great news. We need two things:
- Patch the CMakeLists.txt to output the dll in the bin folder instead of lib. This should be as easy as replacing this line https://github.com/gflags/gflags/blob/v2.2.0/CMakeLists.txt#L504 by another using  RUNTIME DESTINATION bin and LIBRARY DESTINATION lib
- Figure out how to get gflags to build with vc9 and vc10. Adding msintypes as suggested by @jakirkham might do it and/or using [INTTYPES_FORMAT](https://github.com/gflags/gflags/blob/v2.2.0/CMakeLists.txt#L252). I also suspect that https://github.com/gflags/gflags/blob/v2.2.0/CMakeLists.txt#L228 should depend on the VC version.

willyd commented:
If you want to investigate the second point I can fix the first one.

wesm commented:
Why do we put the DLL in the bin folder? cc @maxris

willyd commented:
conda adds this folder to the `PATH` environment variable when activating environment. Otherwise, packages depending on gflags will have to add the lib folder to the `PATH` or copy around the DLL to work properly.

wesm committed:
Patch CMakeLists.txt to install libraries in bindir

wesm committed:
Remove existing post-install patch

wesm committed:
Check in bld.bat changes

jakirkham commented:
I thought `conda` added both `bin` and `lib` to the path on Windows. 😕

willyd commented:
> I thought conda added both bin and lib to the path on Windows. 😕

AFAIK only `bin` is added.

Edit: maybe the `conda` devs should consider adding `lib` too since other libraries install the DLL part to `lib`, boost for example.

jakirkham committed:
Merge remote-tracking branch 'conda-forge/master' into 'wesm/windows'

jakirkham committed:
MNT: Re-rendered with conda-smithy 2.2.2

jakirkham commented on recipe/meta.yaml:
```diff
@@ -7,22 +7,36 @@ package:
+  # skipping
+  # py27 (vc9) and py34 (vc10) since vc does not have stdint.h
+  # py36 since it uses vc14 like py35
+  skip: true  # [win and py!=35]
```
FYI this skip condition would have also skipped `py27`.

Edit: Guessing this was done intentionally because of this [issue]( https://github.com/conda-forge/gflags-feedstock/pull/4#issuecomment-276494303 ).

willyd replied:
You guessed right.

jakirkham committed:
Use NMake to build on Windows

Seem to be having some issues with the Visual Studio generators.
Particular when it comes to building with VS 2008. So this switches over
to NMake, which we have historically had more success with.

jakirkham committed:
Add archive install location to Windows patch

jakirkham commented on recipe/windows_cmake_install_bindir.patch:
```diff
@@ -0,0 +1,29 @@
++               EXPORT ${EXPORT_NAME}
++               RUNTIME DESTINATION ${RUNTIME_INSTALL_DIR}
++               LIBRARY DESTINATION ${LIBRARY_INSTALL_DIR}
++               ARCHIVE DESTINATION ${LIBRARY_INSTALL_DIR})
```
Shouldn't this be `ARCHIVE_INSTALL_DIR`?

willyd replied:
Fixed.

jakirkham commented on recipe/windows_cmake_install_bindir.patch:
```diff
@@ -0,0 +1,29 @@
++               EXPORT ${EXPORT_NAME}
++               RUNTIME DESTINATION ${RUNTIME_INSTALL_DIR}
++               LIBRARY DESTINATION ${LIBRARY_INSTALL_DIR}
++               ARCHIVE DESTINATION ${LIBRARY_INSTALL_DIR})
```
Ditto.

willyd committed:
Install import lirbary in lib and dll in bin

willyd committed:
Added ARCHIVE_INSTALL_DIR and test for import library

jakirkham commented on recipe/windows_cmake_install_bindir.patch:
```diff
@@ -0,0 +1,37 @@
+@@ -399,6 +399,7 @@ endif ()
+ if (OS_WINDOWS)
+   set (RUNTIME_INSTALL_DIR Bin)
+   set (LIBRARY_INSTALL_DIR Lib)
```
Do we still want this one to go to `Bin`?

willyd replied:
Yes otherwise consuming projects won't be able to find the dll.

jakirkham replied:
So are you planning to fix this then?

willyd replied:
Sorry for the misunderstanding I thought the comment was about `RUNTIME_INSTALL_DIR`. The patch is OK the way it is. It is and should be:
```
set (RUNTIME_INSTALL_DIR Bin)
set (LIBRARY_INSTALL_DIR Lib)
set (ARCHIVE_INSTALL_DIR Lib)
```

jakirkham replied:
I thought based on [this]( https://github.com/conda-forge/gflags-feedstock/pull/4#issuecomment-291855570 ) you wanted `LIBRARY_INSTALL_DIR` to be `Bin`. Is this something you don't want now or am I misunderstanding something?

willyd replied:
No. Previously the `install` command was only using `LIBRARY_INSTALL_DIR` so everything was installed in `lib`. Since the `.dll` part of the library is considered a `RUNTIME` we needed to add different destinations for the `RUNTIME`, `LIBRARY` and `ARCHIVE` in the `install` command in order to have the `.dll` in `bin` and keep the `.lib` in `lib`. See https://cmake.org/cmake/help/v3.8/command/install.html#id3 for more details. Clearer now?

jakirkham replied:
Yes, now it is clear. `.dll`s get placed in `RUNTIME_INSTALL_DIR` (`bin` in our case). Thanks for clearing that up for me.

jakirkham replied:
It would be good to raise an issue or do a PR upstream to get this change integrated into `gflags` proper if it is not already. Thanks.

willyd replied:
Right. I'll do that.

willyd commented:
I believe all required changes have been implemented. So apart from a rebase this can be merged in my opinion.

jakirkham commented:
Thanks @willyd and @wesm.

wesm commented:
thanks!

jakirkham commented:
Feel free to submit a patch for static libs, @wesm.
